hkdfchacha20poly1305/hkdfchachapoly.go

```
@@ -0,0 +1,87 @@
+package hkdfchacha20poly1305
+
+import (
+	"crypto/cipher"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"io"
+
+	"golang.org/x/crypto/chacha20poly1305"
+	"golang.org/x/crypto/hkdf"
+)
+
+var sigma = [4]uint32{0x61707865, 0x3320646e, 0x79622d32, 0x6b206574}
+
+type hkdfchacha20poly1305 struct {
+	key [KeySize]byte
+}
+
+const (
+	// KeySize is the size of the key used by this AEAD, in bytes.
+	KeySize = 32
+	// NonceSize is the size of the nonce used with this AEAD, in bytes.
+	NonceSize = 24
+	// HkdfInfo is the parameter used internally for Hkdf's info parameter.
+	HkdfInfo = "TENDERMINT_SECRET_CONNECTION_FRAME_KEY_DERIVE"
+)
+
+//New xChaChapoly1305 AEAD with 24 byte nonces
+func New(key []byte) (cipher.AEAD, error) {
+	if len(key) != KeySize {
+		return nil, errors.New("chacha20poly1305: bad key length")
+	}
+	ret := new(hkdfchacha20poly1305)
+	copy(ret.key[:], key)
+	return ret, nil
+
+}
+func (c *hkdfchacha20poly1305) NonceSize() int {
+	return NonceSize
+}
+
+func (c *hkdfchacha20poly1305) Overhead() int {
+	return 16
+}
+
+func (c *hkdfchacha20poly1305) Seal(dst, nonce, plaintext, additionalData []byte) []byte {
+	if len(nonce) != NonceSize {
+		panic("hkdfchacha20poly1305: bad nonce length passed to Seal")
+	}
+
+	if uint64(len(plaintext)) > (1<<38)-64 {
+		panic("hkdfchacha20poly1305: plaintext too large")
+	}
+
+	var subKey [KeySize]byte
+	var chachaNonce [chacha20poly1305.NonceSize]byte
+
+	hash := sha256.New
+	hkdf := hkdf.New(hash, c.key[:], nonce, []byte(HkdfInfo))
+	io.ReadFull(hkdf, subKey[:])
+	io.ReadFull(hkdf, chachaNonce[:])
+
+	chacha20poly1305, _ := chacha20poly1305.New(subKey[:])
```

Usually bad practice to ignore an error. Add a comment why this can't err here or handle the error.